### PR TITLE
find innercircle of contour by using pointPolygonTest:

### PR DESCRIPTION
**Pull request by jsxyhelu**
Previous PR: #12187

```
force_builders=docs,linux
```

### CODE REVIEW

**Comment by jsxyhelu** [comment]
this is also a good example by using pointPolygonTest
the result is something like:
43351171-099574ca-9241-11e8-9bb5-5f48f349ae11

**Comment by alalek** [comment]
Avoid re-opening of new pullrequests.
Fix build problems "inplace".

**Comment by alalek** [comment]
You didn't push any changes into your fork on source branch "3.4" of this PR.

**Comment by jsxyhelu** [comment]
build successful ,hope pass.

File: samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp [hunk 1]
```diff
@@ -2,15 +2,28 @@
...
+    std::vector<std::vector<cv::Point> >contours;
+    findContours(src,contours,RETR_LIST,CHAIN_APPROX_SIMPLE);
+    for (uint i=0;i<contours.size();i++){
+        double itmp = contourArea(contours[i]);
+        if (imaxcontour < itmp ){
+            icount = i;
+            imaxcontour = itmp;
+        }
+    }
+    return contours[icount];
```
[the remark this answers is not in the document]
Comment (jsxyhelu) [reply samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp]: > This code doesn't handle case if no contours are detected (contours is empty).

you are right,add 

if (0 == contours.size())
	{
		return vector<Point>(NULL);
	}

to handle

File: samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp [hunk 2]
```diff
@@ -2,15 +2,28 @@
...
 #include "opencv2/imgproc.hpp"
 #include <iostream>
-
 using namespace cv;
 using namespace std;
-
+//return the biggest contour by size
+static vector<Point> FindBiggestContour(Mat src){
+    int icount = 0;
+    double imaxcontour = -1;
```
[the remark this answers is not in the document]
Comment (jsxyhelu) [reply samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp]: get it

File: samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp [hunk 3]
```diff
@@ -2,15 +2,28 @@
...
  */
-
 #include "opencv2/highgui.hpp"
 #include "opencv2/imgproc.hpp"
 #include <iostream>
-
 using namespace cv;
 using namespace std;
-
+//return the biggest contour by size
```
[the remark this answers is not in the document]
Comment (jsxyhelu) [reply samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp]: fix it

File: samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp [hunk 4]
```diff
@@ -76,11 +83,28 @@ int main( void )
             }
         }
     }
-
+    //get the biggest Contour
+    vector<Point> biggestContour = FindBiggestContour(src);
```
[the remark this answers is not in the document]
Comment (jsxyhelu) [reply samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp]: have handle biggestContour now

File: samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp [hunk 5]
```diff
@@ -76,11 +83,28 @@ int main( void )
...
+    //find the maximum enclosed circle
+    double dist = 0;
+    double maxdist = 0;
+    Point center;
+    for(int i=0;i<src.cols;i++)
+    {
+        for(int j=0;j<src.rows;j++)
```
[the remark this answers is not in the document]
Comment (jsxyhelu) [reply samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp]: sorry for this

File: samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp [hunk 6]
```diff
@@ -28,17 +40,14 @@ int main( void )
...
     /// Draw it in src
     for( int i = 0; i < 6; i++ )
     {
         line( src, vert[i],  vert[(i+1)%6], Scalar( 255 ), 3 );
     }
-
     /// Get the contours
     vector<vector<Point> > contours;
     findContours( src, contours, RETR_TREE, CHAIN_APPROX_SIMPLE);
-
```
[the remark this answers is not in the document]
Comment (jsxyhelu) [reply samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp]: fix it

File: samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp [hunk 7]
```diff
@@ -48,12 +57,10 @@ int main( void )
             raw_dist.at<float>(i,j) = (float)pointPolygonTest( contours[0], Point2f((float)j, (float)i), true );
         }
     }
-
     double minVal, maxVal;
     minMaxLoc( raw_dist, &minVal, &maxVal );
     minVal = abs(minVal);
     maxVal = abs(maxVal);
-
```
[the remark this answers is not in the document]
Comment (jsxyhelu) [reply samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp]: fix it

File: samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp [hunk 8]
```diff
@@ -76,11 +83,28 @@ int main( void )
             }
         }
     }
-
+    //get the biggest Contour
+    vector<Point> biggestContour = FindBiggestContour(src);
+    //find the maximum enclosed circle
+    double dist = 0;
```
[the remark this answers is not in the document]
Comment (jsxyhelu) [reply samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp]: get it

**Comment by jsxyhelu** [comment]
bug fix,hope pass

File: samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp [hunk 9]
```diff
@@ -77,9 +95,33 @@ int main( void )
         }
     }
 
+    //get the biggest Contour
+    vector<Point> biggestContour = FindBiggestContour(src);
+    //handle case if biggestContour is empty
+    CV_Assert(0 == biggestContour.size());
```
[the remark this answers is not in the document]
Comment (jsxyhelu) [reply samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp]: yes,alalek,and it passed the buildbot.

[the remark this answers is not in the document]
Comment (jsxyhelu) [reply samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp]: default

File: samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp [hunk 10]
```diff
@@ -77,9 +95,33 @@ int main( void )
...
+    //find the maximum enclosed circle
+    double maxdist = 0;
+    Point center;
+    //get the rect bounding the BiggestContour
+    Rect rectBoundingBiggestContour = boundingRect(Mat(biggestContour));
+    for(int i=0;i<rectBoundingBiggestContour.width;i++)
+    {
+        for(int j=0;j<rectBoundingBiggestContour.height;j++)
+        {
+            Point tmpPoint = Point(rectBoundingBiggestContour.x + i,rectBoundingBiggestContour.y + j);
```
[the remark this answers is not in the document]
Comment (jsxyhelu) [reply samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp]: yes,you are right,I will update the code.

File: samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (jsxyhelu) [reply samples/cpp/tutorial_code/ShapeDescriptors/pointPolygonTest_demo.cpp]: yes,you are right,it should be
`CV_Assert(contours.size() != 0);`

**Comment by alalek** [comment]
Related [tutorial](http://pullrequest.opencv.org/buildbot/export/pr/12206/docs/dc/d48/tutorial_point_polygon_test.html) has been wrote for C++ / Python / Java languages.
Need to update other examples too.
Also update [images](https://github.com/opencv/opencv/tree/3.4.2/doc/tutorials/imgproc/shapedescriptors/point_polygon_test/images) in tutorial.

**Comment by jsxyhelu** [comment]
> of course

**Comment by jsxyhelu** [comment]
@alalek 
buildbot passed
wait admininstrator pass

File: samples/java/tutorial_code/ShapeDescriptors/point_polygon_test/PointPolygonTestDemo.java [hunk 11]
```diff
@@ -86,8 +84,6 @@ public void run() {
     public static void main(String[] args) {
         // Load the native OpenCV library
         System.loadLibrary(Core.NATIVE_LIBRARY_NAME);
-
         new PointPolygonTest().run();
     }
-
```
Comment (alalek) [inline samples/java/tutorial_code/ShapeDescriptors/point_polygon_test/PointPolygonTestDemo.java]: Revert unnecessary changes (removal of empty lines).

> HighGui.imshow("Distance", drawing);

Need to update the name in Java sample too too.

File: samples/python/tutorial_code/ShapeDescriptors/point_polygon_test/pointPolygonTest_demo.py [hunk 12]
```diff
@@ -46,6 +46,7 @@
             drawing[i,j,1] = 255
             drawing[i,j,2] = 255
 
+cv.circle(drawing,maxDistPt, np.int(maxVal),(255,255,255), 1, cv.LINE_8, 0)
```
Comment (alalek) [inline samples/python/tutorial_code/ShapeDescriptors/point_polygon_test/pointPolygonTest_demo.py]: Respect the current coding style of the file and add space after each comma `,` (except may be tuple `(255,255,255)`).

> np.int(maxVal)

Just `int(maxVal)` should work too.

Comment (jsxyhelu) [reply samples/python/tutorial_code/ShapeDescriptors/point_polygon_test/pointPolygonTest_demo.py]: great suggestion!I will modifiy the code as soom as possible

**Comment by jsxyhelu** [comment]
I will recover "Point_Polygon_Test_Result.jpg" ,and avoid to change markdown.

**Comment by jsxyhelu** [comment]
@alalek
code changed
wait admininstrator pass
